File: spec/models/character_spec.rb
require 'spec_helper'

describe Character do
  let (:new_character) { Factory :character }
  let (:new_equipped_public_asset) {GameAsset.create(equipped: true, legality: "Public",name: "1", price: 500, character: new_character, amount: 500) }
  let (:new_equipped_restricted_asset) {GameAsset.create(equipped: true, legality: "Restricted",name: "2", price: 500, character: new_character, amount: 500)}
  let (:new_equipped_forbidden_asset) {GameAsset.create(equipped: true, legality: "Forbidden",name: "3", price: 500, character: new_character, amount: 500) }
  let (:new_unequipped_public_asset) {GameAsset.create(equipped: false, legality: "Public",name: "1", price: 500, character: new_character, amount: 500)  }
  let (:new_unequipped_restricted_asset) {GameAsset.create(equipped: false, legality: "Restricted",name: "2", price: 500, character: new_character, amount: 500)}
  let (:new_unequipped_forbidden_asset) {GameAsset.create(equipped: false, legality: "Forbidden",name: "3", price: 500, character: new_character, amount: 500)  }

  describe "model associations" do
    it "has one commlink" do
      commlink = Factory :commlink, character_id: new_character.id
      new_character.commlink.should == commlink
    end
  end

  describe "checking legality status method" do

    it "should only check equipped assets when equipped asset is on top" do
      new_equipped_public_asset
      new_unequipped_forbidden_asset

      new_character.check_legality_status.should == "Public"
    end

    it "should only check equipped assets when equipped is asset is at bottom" do
      new_unequipped_forbidden_asset
      new_equipped_public_asset

      new_character.check_legality_status.should == "Public"
    end

    it "should see that character status is Forbidden" do
      new_equipped_public_asset
      new_unequipped_public_asset
      new_equipped_restricted_asset
      new_unequipped_restricted_asset
      new_equipped_forbidden_asset
      new_unequipped_forbidden_asset
      new_equipped_public_asset

      new_character.check_legality_status.should == "Forbidden"
    end

    it "should see that character status is Restricted" do
      new_equipped_public_asset
      new_unequipped_public_asset
      new_equipped_restricted_asset
      new_unequipped_forbidden_asset
      new_equipped_public_asset

      new_character.check_legality_status.should == "Restricted"
    end

    it "should see that character status is Public" do
      new_equipped_public_asset
      new_unequipped_restricted_asset
      new_equipped_public_asset

      new_character.check_legality_status.should == "Public"
    end
  end

  describe " cheking sum_a method " do

    it "should add all the assets up" do
      new_equipped_public_asset
      new_equipped_restricted_asset
      new_equipped_forbidden_asset
      new_unequipped_public_asset
      new_unequipped_restricted_asset
      new_unequipped_forbidden_asset

      new_character.sum_a.should == 1500000
    end

    it "should add all the expenses up" do
      GameExpense.create(name: "Rent2", price: 500, character: new_character)
      GameExpense.create(name: "Rent", price: 500, character: new_character )

      new_character.sum_e.should == 1000
    end

  end
end

File: spec/javascripts/models/character_spec.js
(function() {

  describe("Character", function() {
    beforeEach(function() {
      GameAsset.collection.reset();
      return GameExpense.collection.reset();
    });
    describe("assets", function() {
      it("includes assets matching id", function() {
        var asset, character;
        character = new Character({
          id: 1
        });
        asset = new GameAsset({
          character_id: character.id
        });
        GameAsset.collection.add(asset);
        return expect(character.assets()).toContain(asset);
      });
      return it("doesn't include asset for another character", function() {
        var asset, character;
        character = new Character({
          id: 1
        });
        asset = new GameAsset({
          character_id: 2
        });
        GameAsset.collection.add(asset);
        return expect(character.assets()).not.toContain(asset);
      });
    });
    describe("assetTotal", function() {
      return it("is total of character's assets", function() {
        var character;
        character = new Character({
          id: 1
        });
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          price: 2,
          amount: 3
        }));
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          price: 4,
          amount: 2
        }));
        return expect(character.assetTotal()).toBe(14);
      });
    });
    describe("expenses", function() {
      it("includes expenses matching id", function() {
        var character, expense;
        character = new Character({
          id: 1
        });
        expense = new GameExpense({
          character_id: character.id
        });
        GameExpense.collection.add(expense);
        return expect(character.expenses()).toContain(expense);
      });
      return it("doesn't include expense for another character", function() {
        var character, expense;
        character = new Character({
          id: 1
        });
        expense = new GameExpense({
          character_id: 2
        });
        GameExpense.collection.add(expense);
        return expect(character.expenses()).not.toContain(expense);
      });
    });
    describe("expenseTotal", function() {
      return it("is total of character's expenses", function() {
        var character;
        character = new Character({
          id: 1
        });
        GameExpense.collection.add(new GameExpense({
          character_id: character.id,
          price: 2
        }));
        GameExpense.collection.add(new GameExpense({
          character_id: character.id,
          price: 4
        }));
        return expect(character.expenseTotal()).toBe(6);
      });
    });
    return describe("legalStatus", function() {
      it("is public when character has no assets", function() {
        var character;
        character = new Character({
          id: 1
        });
        return expect(character.legalStatus()).toBe("Public");
      });
      it("is restricted if restricted asset is equipped", function() {
        var character;
        character = new Character({
          id: 1
        });
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          equipped: true,
          legality: 'Restricted'
        }));
        return expect(character.legalStatus()).toBe("Restricted");
      });
      it("is forbidden if forbidden asset is equipped", function() {
        var character;
        character = new Character({
          id: 1
        });
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          equipped: true,
          legality: 'Forbidden'
        }));
        return expect(character.legalStatus()).toBe("Forbidden");
      });
      it("is forbidden if borth resticted and forbidden equipped", function() {
        var character;
        character = new Character({
          id: 1
        });
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          equipped: true,
          legality: 'Restricted'
        }));
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          equipped: true,
          legality: 'Forbidden'
        }));
        return expect(character.legalStatus()).toBe("Forbidden");
      });
      return it("is public if forbidden assets are not equipped", function() {
        var character;
        character = new Character({
          id: 1
        });
        GameAsset.collection.add(new GameAsset({
          character_id: character.id,
          equipped: false,
          legality: 'Forbidden'
        }));
        return expect(character.legalStatus()).toBe("Public");
      });
    });
  });

}).call(this);
